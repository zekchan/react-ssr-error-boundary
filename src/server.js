import React, { Component } from 'react'
import { renderToStaticMarkup } from 'react-dom/server'

// Array to store React contexts
const contexts = []

// Use specific context
export function useContext (Context) {
  contexts.push(Context);
}

// Clear contexts in use
export function clearContexts () {
  contexts.length = 0
}

// Shim `React.createContext()` to capture contexts
let shimmed = false

export function shim () {
  if (shimmed) return
  shimmed = true

  const { createContext } = React
  React.createContext = function() {
    const Context = createContext.apply(this, arguments)
    contexts.push(Context)
    return Context
  }
}

// Create Provider component for specific context types
export function _makeProvider (contextTypes) {
  return class ProvideContext extends Component {
    static childContextTypes = contextTypes

    getChildContext () {
      return this.props.context
    }

    render () {
      return this.props.children
    }
  }
}

// Server-side render, propagating contexts to children
export function _render (self, ProvideContext) {
  // Wrap element in provider for legacy context
  const element = (
    <ProvideContext context={self.context}>
      {self.props.children}
    </ProvideContext>
  )

  // Wrap element in consumers and providers for new context.
  // Consumers are added outside HTML render, to capture value of all contexts.
  // Providers are added inside HTML render, to reinject context values back
  // into the rendered element.
  return stackConsumers(values => {
    const elementWithProviders = stackProviders(element, values)

    try {
      const __html = renderToStaticMarkup(elementWithProviders)
      return <div dangerouslySetInnerHTML={{__html}} />
    } catch (e) {
      return <>{self.props.fallBack()}</>
    }
  })
}

// Create stack of consumer components
function stackConsumers (retFn) {
  const values = []
  let fn = () => retFn(values)

  for (let i = 0; i < contexts.length; i++) {
    const Context = contexts[i]
    const childFn = fn

    fn = () => (
      <Context.Consumer>
        {value => {
          values[i] = value
          return childFn()
        }}
      </Context.Consumer>
    )
  }

  return fn()
}

// Create stack of provider components
function stackProviders (element, values) {
  for (let i = 0; i < values.length; i++) {
    const Context = contexts[i]
    const childElement = element

    element = (
      <Context.Provider value={values[i]}>
        {childElement}
      </Context.Provider>
    )
  }

  return element
}
