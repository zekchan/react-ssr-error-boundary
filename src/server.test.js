/**
 * @jest-environment node
 */
import React from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import ErrorFallback, { withContext } from './index'
import { shim, clearContexts } from './server'
shim()

function FallBack () {
  return <div>FallBack!</div>
}

const errorFn = console.error

function turnOffErrors () {
  console.error = () => {}
}

function turnOnErrors () {
  console.error = errorFn
}

describe('Server side', () => {
  it('Renders child component without errors', () => {
    function GoodComponent () {
      return <div>No errors!</div>
    }

    const html = renderToStaticMarkup(<ErrorFallback fallBack={() => <FallBack />}>
      <GoodComponent />
    </ErrorFallback>)

    expect(html).toBe('<div><div>No errors!</div></div>')
    expect(html).toMatchSnapshot()
  })

  it('Renders fallBack component if children rendering throws error', () => {
    function BadComponent () {
      throw new Error()
    }

    turnOffErrors()
    const html = renderToStaticMarkup(<ErrorFallback fallBack={() => <FallBack />}>
      <BadComponent />
    </ErrorFallback>)

    expect(html).toBe('<div>FallBack!</div>')
    expect(html).toMatchSnapshot()
    turnOnErrors()
  })

  it('Renders child component with context dependencies', () => {
    function GoodComponent (props, context) {
      return <div>No errors! {context.someContext}</div>
    }

    class ContextProvider extends React.Component {
      getChildContext () {
        return {someContext: 'Context variable'}
      }

      render () {
        return this.props.children
      }
    }

    GoodComponent.contextTypes = {
      someContext: () => {}
    }
    ContextProvider.childContextTypes = GoodComponent.contextTypes

    const ErrorFallBackWithContext = withContext(GoodComponent.contextTypes)
    const html = renderToStaticMarkup(
      <ContextProvider>
        <ErrorFallBackWithContext fallBack={() => <FallBack />}>
          <GoodComponent />
        </ErrorFallBackWithContext>
      </ContextProvider>
    )

    expect(html).toBe('<div><div>No errors! Context variable</div></div>')
    expect(html).toMatchSnapshot()
  })

  it('Renders child component with new context dependencies', () => {
    const Context = React.createContext()

    function GoodComponent () {
      return (
        <Context.Consumer>
          {context => <div>No errors! {context.someContext}</div>}
        </Context.Consumer>
      )
    }

    function ContextProvider (props) {
      return (
        <Context.Provider value={{someContext: 'Context variable'}}>
          {props.children}
        </Context.Provider>
      )
    }

    const html = renderToStaticMarkup(
      <ContextProvider>
        <ErrorFallback fallBack={() => <FallBack />}>
          <GoodComponent />
        </ErrorFallback>
      </ContextProvider>
    )

    expect(html).toBe('<div><div>No errors! Context variable</div></div>')
    clearContexts()
  })

  it('Renders child component with multiple new context dependencies', () => {
    const Context1 = React.createContext()
    const Context2 = React.createContext()
    const Context3 = React.createContext()

    function GoodComponent () {
      return (
        <Context1.Consumer>
          {context1 => (
              <Context2.Consumer>
                {context2 => (
                    <Context3.Consumer>
                      {context3 => (
                          <div>No errors! {context1.someContext} {context2.someContext} {context3.someContext}</div>
                      )}
                    </Context3.Consumer>
                )}
              </Context2.Consumer>
          )}
        </Context1.Consumer>
      )
    }

    function ContextProvider (props) {
      return (
        <Context1.Provider value={{someContext: 'Context variable1'}}>
          <Context2.Provider value={{someContext: 'Context variable2'}}>
            <Context3.Provider value={{someContext: 'Context variable3'}}>
              {props.children}
            </Context3.Provider>
          </Context2.Provider>
        </Context1.Provider>
      )
    }

    const html = renderToStaticMarkup(
      <ContextProvider>
        <ErrorFallback fallBack={() => <FallBack />}>
          <GoodComponent />
        </ErrorFallback>
      </ContextProvider>
    )

    expect(html).toBe('<div><div>No errors! Context variable1 Context variable2 Context variable3</div></div>')
    expect(html).toMatchSnapshot()
    clearContexts()
  })

  it('Renders fallBack component if children rendering throws error with new contexts', () => {
    const Context = React.createContext()

    function BadComponent () {
      return (
        <Context.Consumer>
          {() => <BadComponentInner/>}
        </Context.Consumer>
      )
    }

    function BadComponentInner () {
      throw new Error()
    }

    function ContextProvider (props) {
      return (
        <Context.Provider value={{someContext: 'Context variable'}}>
          {props.children}
        </Context.Provider>
      )
    }

    turnOffErrors()
    const html = renderToStaticMarkup(
      <ContextProvider>
        <ErrorFallback fallBack={() => <FallBack />}>
          <BadComponent />
        </ErrorFallback>
      </ContextProvider>
    )

    expect(html).toBe('<div>FallBack!</div>')
    expect(html).toMatchSnapshot()
    turnOnErrors()
    clearContexts()
  })

  it('Running createContext shim twice has no effect', () => {
    const {createContext} = React
    shim()
    expect(React.createContext).toBe(createContext)
  })
})
