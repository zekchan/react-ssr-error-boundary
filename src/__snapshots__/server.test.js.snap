// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server side Renders child component with context dependencies 1`] = `"<div><div>No errors! Context variable</div></div>"`;

exports[`Server side Renders child component with multiple new context dependencies 1`] = `"<div><div>No errors! Context variable1 Context variable2 Context variable3</div></div>"`;

exports[`Server side Renders child component without errors 1`] = `"<div><div>No errors!</div></div>"`;

exports[`Server side Renders fallBack component if children rendering throws error 1`] = `"<div>FallBack!</div>"`;

exports[`Server side Renders fallBack component if children rendering throws error with new contexts 1`] = `"<div>FallBack!</div>"`;
